QRdecomposition fails for some matrices

```
Hello,

on the current git version of sympy I got a strange failure of the
Matrix.QRdecomposition method:

===========================================================================
In [1]: A = Matrix([[1,1],[1,1.01]])

In [2]: A.QRdecomposition()
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)

/Users/voss/sympy/<ipython console> in <module>()

/Users/voss/sympy/sympy/matrices/matrices.py in QRdecomposition(self)
    966             R[j,j] = tmp.norm()
    967             Q[:,j] = tmp / R[j,j]
--> 968             assert Q[:,j].norm() == 1
    969             for i in range(j):
    970                 R[i,j] = Q[:,i].dot(self[:,j])

AssertionError: 
===========================================================================

This seems not to be a problem with the conditioning of A, since it works
in the following case (which has a worse conditioning):

===========================================================================
In [3]: A = Matrix([[1,1],[1,1.001]])

In [4]: A.QRdecomposition()
Out[4]: 
⎛⎡  ⎽⎽⎽                    ⎤                              ⎞
⎜⎢╲╱ 2                     ⎥  ⎡  ⎽⎽⎽              ⎽⎽⎽    ⎤⎟
⎜⎢─────  -0.707106781186548⎥, ⎢╲╱ 2      1.0005⋅╲╱ 2     ⎥⎟
⎜⎢  2                      ⎥  ⎢                          ⎥⎟
⎜⎢                         ⎥  ⎣  0    0.00070710678118647⎦⎟
⎜⎢  ⎽⎽⎽                    ⎥                              ⎟
⎜⎢╲╱ 2                     ⎥                              ⎟
⎜⎢─────  0.707106781186548 ⎥                              ⎟
⎝⎣  2                      ⎦                              ⎠
===========================================================================

I hope this helps,
Jochen
```

Original issue for #4520: http://code.google.com/p/sympy/issues/detail?id=1421
Original author: https://code.google.com/u/jochen.voss@gmail.com/
Original owner: https://code.google.com/u/jochen.voss@gmail.com/

--- a/sympy/matrices/tests/test_decompositions.py
+++ b/sympy/matrices/tests/test_decompositions.py
@@ -3,6 +3,8 @@ from sympy.matrices.matrices import NonSquareMatrixError
 from sympy.matrices import Matrix, zeros, eye, SparseMatrix
 from sympy.abc import x, y, z
 from sympy.testing.pytest import raises
+from sympy.testing.matrices import allclose
+
 
 def test_LUdecomp():
     testmat = Matrix([[0, 2, 5, 3],
@@ -216,28 +218,18 @@ def test_QR_trivial():
     assert A == Q*R
 
 
-def _numerically_equal(A, B, tol=1e-15):
-    if A.shape != B.shape:
-        return False
-
-    for a, b in zip(A, B):
-        if abs(a - b) > tol:
-            return False
-    return True
-
-
 def test_QR_float():
     A = Matrix([[1, 1], [1, 1.01]])
     Q, R = A.QRdecomposition()
-    assert _numerically_equal(Q * R, A)
-    assert _numerically_equal(Q * Q.T, Matrix.eye(2), tol=1e-13)
-    assert _numerically_equal(Q.T * Q, Matrix.eye(2), tol=1e-13)
+    assert allclose(Q * R, A)
+    assert allclose(Q * Q.T, Matrix.eye(2))
+    assert allclose(Q.T * Q, Matrix.eye(2))
 
     A = Matrix([[1, 1], [1, 1.001]])
     Q, R = A.QRdecomposition()
-    assert _numerically_equal(Q * R, A)
-    assert _numerically_equal(Q * Q.T, Matrix.eye(2), tol=1e-13)
-    assert _numerically_equal(Q.T * Q, Matrix.eye(2), tol=1e-13)
+    assert allclose(Q * R, A)
+    assert allclose(Q * Q.T, Matrix.eye(2))
+    assert allclose(Q.T * Q, Matrix.eye(2))
 
 
 def test_LUdecomposition_Simple_iszerofunc():

--- a/sympy/matrices/tests/test_decompositions.py
+++ b/sympy/matrices/tests/test_decompositions.py
@@ -215,6 +215,31 @@ def test_QR_trivial():
     assert R.is_upper
     assert A == Q*R
 
+
+def _numerically_equal(A, B, tol=1e-15):
+    if A.shape != B.shape:
+        return False
+
+    for a, b in zip(A, B):
+        if abs(a - b) > tol:
+            return False
+    return True
+
+
+def test_QR_float():
+    A = Matrix([[1, 1], [1, 1.01]])
+    Q, R = A.QRdecomposition()
+    assert _numerically_equal(Q * R, A)
+    assert _numerically_equal(Q * Q.T, Matrix.eye(2), tol=1e-13)
+    assert _numerically_equal(Q.T * Q, Matrix.eye(2), tol=1e-13)
+
+    A = Matrix([[1, 1], [1, 1.001]])
+    Q, R = A.QRdecomposition()
+    assert _numerically_equal(Q * R, A)
+    assert _numerically_equal(Q * Q.T, Matrix.eye(2), tol=1e-13)
+    assert _numerically_equal(Q.T * Q, Matrix.eye(2), tol=1e-13)
+
+
 def test_LUdecomposition_Simple_iszerofunc():
     # Test if callable passed to matrices.LUdecomposition_Simple() as iszerofunc keyword argument is used inside
     # matrices.LUdecomposition_Simple()

--- a/sympy/testing/matrices.py
+++ b/sympy/testing/matrices.py
@@ -1,0 +1,8 @@
+def allclose(A, B, rtol=1e-05, atol=1e-08):
+    if len(A) != len(B):
+        return False
+
+    for x, y in zip(A, B):
+        if abs(x-y) > atol + rtol * max(abs(x), abs(y)):
+            return False
+    return True

